src/libtls/tls_hkdf.c

```diff
@@ -0,0 +1,520 @@
+/*
+ * Copyright (C) 2020 Pascal Knecht
+ * Copyright (C) 2020 Méline Sieber
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.  See <http://www.fsf.org/copyleft/gpl.txt>.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ */
+
+#include "tls_hkdf.h"
+#include <utils/chunk.h>
+#include <library.h>
+#include <crypto/prf_plus.h>
+
+typedef struct private_tls_hkdf_t private_tls_hkdf_t;
+
+typedef enum hkdf_phase {
+	HKDF_PHASE_0,
+	HKDF_PHASE_1,
+	HKDF_PHASE_2,
+	HKDF_PHASE_3,
+} hkdf_phase;
+
+struct private_tls_hkdf_t {
+
+	/**
+	 * Public tls_hkdf_t interface.
+	 */
+	struct tls_hkdf_t public;
+
+	/**
+	 * Phase we are in.
+	 */
+	hkdf_phase phase;
+
+	/**
+	 * Hash algorithm used.
+	 */
+	hash_algorithm_t hash_algorithm;
+
+	/**
+	 * Length of this hash output in bytes.
+	 */
+	size_t hash_length;
+
+	/**
+	 * Preshared key IKM
+	 */
+	chunk_t *psk;
+
+	/**
+	 * (EC)DHE IKM
+	 */
+	chunk_t *shared_secret;
+
+	/**
+	 * Pseudorandom function used.
+	 */
+	prf_t *prf;
+
+	/**
+	 * Pseudorandom plus function used.
+	 */
+	prf_plus_t *prf_plus;
+
+	/**
+	 * Hasher used.
+	 */
+	hasher_t *hasher;
+
+	/**
+	 * Salt used.
+	 */
+	chunk_t salt;
+
+	/**
+	 * IKM used.
+	 */
+	chunk_t ikm;
+
+	/**
+	 * PRK used.
+	 */
+	chunk_t prk;
+
+	/**
+	 * Info used.
+	 */
+	chunk_t info;
+
+	/**
+	 * OKM used.
+	 */
+	chunk_t okm;
+
+	/**
+	 * Length of the desired key output.
+	 */
+	size_t L;
+};
+
+static char *hkdf_labels[] = {
+	"tls13 ext binder",
+	"tls13 res binder",
+	"tls13 c e traffic",
+	"tls13 e expt master",
+	"tls13 c hs traffic",
+	"tls13 s hs traffic",
+	"tls13 c ap traffic",
+	"tls13 s ap traffic",
+	"tls13 exp master",
+	"tls13 res master",
+};
+
+/* HKDF-Extract(this->salt, this->ikm) -> this->prk */
```

Comment:
No need for `this->` here, just write it as it is in [RFC 5869](https://tools.ietf.org/html/rfc5869) and maybe even reference the RFC, that is, write something like:
```suggestion
/**
 * Step 1: Extract, as defined in RFC 5869, section 2.2:
 * HKDF-Extract(salt, IKM) -> PRK
 */